File: Online-judge/leetcode/predict-the-winner.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/predict-the-winner/editorial/
 * Time complexity: O(N^2).
 * Space complexity: O(N^2).
 */

class Solution {
private:
    vector<vector<int>> dp;

    int solve(vector<int>& nums, int left, int right) {
        if (dp[left][right] != -1) return dp[left][right];
        if (left == right) {
            return nums[left];
        }
        int leftScore = nums[left] - solve(nums, left + 1, right);
        int rightScore = nums[right] - solve(nums, left, right - 1);

        dp[left][right] = max(leftScore, rightScore);
        return dp[left][right];
    }
    
public:
    bool PredictTheWinner(vector<int>& nums) {
        int n = nums.size();
        dp = vector<vector<int>>(n, vector<int>(n, -1));
        return solve(nums, 0, nums.size() - 1) >= 0;
    }
};

File: Online-judge/leetcode/predict-the-winner.py
from typing import List


class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [[0] * n for _ in range(n)]

        for index in range(n):
            dp[index][index] = nums[index] 

        for index in range(1, n):
            for left in range(n - index):
                right = left + index
                dp[left][right] = max(nums[left] - dp[left + 1][right], nums[right] - dp[left][right - 1])
                
        return dp[0][n - 1] >= 0